terraform/node_module_removed.go

```diff
@@ -0,0 +1,71 @@
+package terraform
+
+import (
+	"fmt"
+	"log"
+	"reflect"
+)
+
+// NodeModuleRemoved represents a module that is no longer in the
+// config.
+type NodeModuleRemoved struct {
+	PathValue []string
+}
+
+func (n *NodeModuleRemoved) Name() string {
+	return fmt.Sprintf("%s (removed)", modulePrefixStr(n.PathValue))
+}
+
+// GraphNodeSubPath
+func (n *NodeModuleRemoved) Path() []string {
+	return n.PathValue
+}
+
+// GraphNodeEvalable
+func (n *NodeModuleRemoved) EvalTree() EvalNode {
+	return &EvalOpFilter{
+		Ops: []walkOperation{walkRefresh, walkApply, walkDestroy},
+		Node: &EvalDeleteModule{
+			PathValue: n.PathValue,
+		},
+	}
+}
+
+// EvalDeleteModule is an EvalNode implementation that removes an empty module
+// entry from the state.
+type EvalDeleteModule struct {
+	PathValue []string
+}
+
+func (n *EvalDeleteModule) Eval(ctx EvalContext) (interface{}, error) {
+	state, lock := ctx.State()
+	if state == nil {
+		return nil, nil
+	}
+
+	// Get a write lock so we can access this instance
+	lock.Lock()
+	defer lock.Unlock()
+
+	// Make sure we have a clean state
+	// Destroyed resources aren't deleted, they're written with an ID of "".
```

Comment:
:thinking: this seems like a bug in itself! But indeed probably best saved to fix another day...

Reply:
It seems to have been done purposely, and the resource is removed by `prune`. THat's probably best for a later cleanup.

Reply:
This is definitely an issue to research later. 
Pruning the state after every EvalWriteState causes a large number of interpolation errors in tests, where the resource disappears before it can be interpolated. That could totally be from incorrect test fixtures though, but I'm going to postpone cleaning that up until after 0.11.